### PR TITLE
[SPARK-27680][CORE][SQL][GRAPHX] Remove usage of Traversable

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

This removes usage of `Traversable`, which is removed in Scala 2.13. This is mostly an internal change, except for the change in the `SparkConf.setAll` method. See additional comments below.

## How was this patch tested?

Existing tests.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 1]
```diff
@@ -168,6 +168,15 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
   }
 
   /** Set multiple parameters together */
+  def setAll(settings: Iterable[(String, String)]): SparkConf = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: Any caller is almost certainly passing an object that is `Iterable` as well as `Traversable`, but this maintains binary compatibility for now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/javaCode.scala [hunk 2]
```diff
@@ -183,7 +183,7 @@ trait Block extends TreeNode[Block] with JavaCode {
     def doTransform(arg: Any): AnyRef = arg match {
       case e: ExprValue => transform(e)
       case Some(value) => Some(doTransform(value))
-      case seq: Traversable[_] => seq.map(doTransform)
+      case seq: Iterable[_] => seq.map(doTransform)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/javaCode.scala]: This is the only part of the change that is of concern. It's internal, but this now only matches `Iterable` and not `Traversable`, which are not the same in Scala 2.12. However, AFAICT, no standard class we'd expect here doesn't extend `Iterable`. See https://docs.scala-lang.org/overviews/collections/overview.html ; `Seq`, `Set` and `Map` are also `Iterable` already. This shouldn't change behavior, but tests will help prove that.

**Comment by beliefer** [comment]
The internal change is easy to understand.

**Review by felixcheung — APPROVED** [review]

**Comment by srowen** [comment]
BTW we have a bigger related issue: `TraversableOnce` -> `IterableOnce`. This is much more widely used and in some key APIs. I may break that one out separately into a new JIRA.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 3]
```diff
@@ -168,6 +168,15 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
...
+    settings.foreach { case (k, v) => set(k, v) }
+    this
+  }
+
+  /**
+   * Set multiple parameters together
+   */
+  @deprecated("Use setAll(Iterable) instead", "3.0.0")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: Hm, is it even worth keeping and deprecating this and removing it later? It'll be source compatible already and binary compatibility isn't guaranteed; one would have to go out of one's way to call the deprecated method at all, with a cast or with some odd custom collection

Comment (srowen) [reply core/src/main/scala/org/apache/spark/SparkConf.scala]: One reason I put this forth is that the `TraversableOnce` problem is harder; its replacement `IterableOnce` doesn't exist in Scala 2.12. The same overload strategy doesn't work out quite right for the places it's used (generic types and function signatures make the overload ambiguous). I am considering proposing just changing the type there for 3.0.

Comment (felixcheung) [reply core/src/main/scala/org/apache/spark/SparkConf.scala]: I can see setAll() is being used though. but maybe it won't work even after a rebuild (bin compat)?

Comment (srowen) [reply core/src/main/scala/org/apache/spark/SparkConf.scala]: Calls to `setAll` will now bind to the more specific `Iterable` arg version in all cases in Spark, and I'm not aware of any Scala classes that aren't `Iterable` to begin with. One consequence I suppose is that passing an `Iterator` will use the deprecated method. But it'll still be there. I don't anticipate we'll remove it until we want to support Scala 2.13 and that won't be in 3.0. For now, no method is removed so should be no problem.

**Review by HyukjinKwon — APPROVED** [review]

**Review by gaborgsomogyi — APPROVED** [review]
LGTM.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**